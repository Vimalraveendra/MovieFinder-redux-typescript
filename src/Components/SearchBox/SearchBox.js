import React from "react";
import styles from "./SearchBox.module.css";
// import { connect } from "react-redux";
// import {
//   searchChange,
//   clearSearchField,
// } from "../../Redux/SearchBox/SearchBox.actions";
// import {
//   fetchFilmDataStart,
//   clearMovies,
// } from "../../Redux/FilmCard/FilmCard.actions";

// import { createStructuredSelector } from "reselect";
// import { selectError } from "../../Redux/FilmCard/FilmCard.selectors";
// import { selectSearchField } from "../../Redux/SearchBox/SearchBox.selectors";

// import { SearchBoxActionTypes } from "../../Redux/SearchBox/SearchBox.types";
// import { FilmCardActionTypes } from "../../Redux/FilmCard/FilmCard.types";

// // import { AppState } from "../../Redux/store";

// // //here we are doing aggregation of all the app action types
// // // & use it inside the mapDispatchToProps to dispatch actions.
// // type AppDispatchActions = SearchBoxActionTypes | FilmCardActionTypes;

// // here we are going to specify the type of the parameters that
// // SearchBox components have
// // interface ISearchBox {
// //   searchChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
// //   clearMovies: () => void;
// //   handleSubmit: (searchField: string) => void;
// //   clearSearchField: () => void;
// //   searchField: string;
// //   error: string;
// // }

const SearchBox = ({
  searchChange,
  handleSubmit,
  clearMovies,
  searchField,
  error,
  clearSearchField,
}) => {
  return (
    <div className={styles.container}>
      <input
        className={styles.input}
        type="search"
        placeholder="search movies here"
        onChange={searchChange}
        value={searchField}
      />

      <button
        className={styles.button}
        onClick={() => handleSubmit(searchField)}
      >
        Search
      </button>
      {searchField.length > 0 || error ? (
        <span
          className={styles.icon}
          role="img"
          aria-label="sheep"
          onClick={() => {
            clearMovies();
            clearSearchField();
          }}
        >
          &#10060;
        </span>
      ) : null}
    </div>
  );
};

// interface LinKStateProps {
//   searchField: string;
//   error : string;
// }

// here we are specifying the return type of LinkMapDispatchProps
// interface LinkMapProps {
//   //  here we are dispatching actions to redux store so that we are not
//   //  returning anything that why we mentioned void
//   searchChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
//   clearMovies: () => void;
//   handleSubmit: (searchField: string) => void;
//   clearSearchField: () => void;
// }
// const mapStateToProps = createStructuredSelector<AppState, LinKStateProps>({
//   searchField: selectSearchField,
//   error: selectError,
// });

// const mapDispatchToProps = (
//   // here we are going to specify the type of the total app actions that are going to dispatch to
//   // redux store using the Dispatch keyword from redux & type AppDispatchActions
//   // then we are going to specify the return type of all these total app actions
//   // using the interface LinkMapProps
//   dispatch: Dispatch<AppDispatchActions>
// ): LinkMapProps => ({
//   searchChange: (event) => dispatch(searchChange(event.target.value)),
//   clearMovies: () => dispatch(clearMovies()),
//   handleSubmit: (searchField) => dispatch(fetchFilmDataStart(searchField)),
//   // when the user click on the close icon we need to make the
//   // input text empty
//   clearSearchField: () => dispatch(clearSearchField()),
// });

export default SearchBox;
