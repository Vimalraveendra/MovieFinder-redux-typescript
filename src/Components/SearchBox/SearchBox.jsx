import React from "react";
import styles from "./SearchBox.module.css";
import { connect } from "react-redux";
import {
  searchChange,
  clearSearchField,
} from "../../Redux/SearchBox/SearchBox.actions";
import {
  fetchedFilmData,
  clearMovies,
} from "../../Redux/FilmCard/FilmCard.actions";

const SearchBox = ({
  searchChange,
  handleSubmit,
  clearMovies,
  searchField,
  error,
  clearSearchField,
}) => {
  console.log("search", searchField);
  return (
    <div className={styles.container}>
      <input
        className={styles.input}
        type="search"
        placeholder="search movies here"
        onChange={searchChange}
        value={searchField}
      />

      <button
        className={styles.button}
        onClick={() => handleSubmit(searchField)}
      >
        Search
      </button>
      {searchField.length > 0 || error ? (
        <span
          className={styles.icon}
          role="img"
          aria-label="sheep"
          onClick={() => {
            clearMovies();
            clearSearchField();
          }}
        >
          &#10060;
        </span>
      ) : null}
    </div>
  );
};

const mapStateToProps = ({
  moviesList: { error },
  searchText: { searchField },
}) => ({
  searchField,
  error,
});

const mapDispatchToProps = (dispatch) => ({
  searchChange: (e) => dispatch(searchChange(e.target.value)),
  clearMovies: () => dispatch(clearMovies()),
  handleSubmit: (searchField) => dispatch(fetchedFilmData(searchField)),
  // when the user click on the close icon we need to make the
  // input text empty
  clearSearchField: () => dispatch(clearSearchField()),
});

export default connect(mapStateToProps, mapDispatchToProps)(SearchBox);
