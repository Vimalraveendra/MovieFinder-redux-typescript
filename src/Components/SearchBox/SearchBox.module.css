.container {
  /* causes absolute icon span to be positioned correctly */
  display: inline-block;
  position: relative;
  width: 20rem;
  height: 3.1rem;
  box-sizing: border-box;
}

.input {
  /* background: #bdbdbdfa; */
  width: 100%;
  height: 100%;

  border-radius: 5px;
  border: none;
  text-transform: capitalize;
  padding-left: 3.5rem;
  font-size: 1rem;
}
.input:focus {
  outline: none;
}

.button {
  /* These are set relative to the height of the input 
box to bound the box neatly inside. This is aesthetic 
to me but you can change the dimensions of course. */

  position: absolute;
  right: 0.2rem;
  top: 0.15rem;
  width: 4.6rem;
  height: 2.8rem;
  border-radius: 3px;
  background-color: rgba(0, 255, 0, 0.7);
  font-size: 1rem;
  outline: none;
  box-sizing: border-box;
}

.icon {
  position: absolute;
  left: 0.2rem;
  top: 0.5rem;

  font-size: 1.8rem;
}
