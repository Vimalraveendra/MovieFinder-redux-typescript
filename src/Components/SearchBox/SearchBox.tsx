import React from "react";
import styles from "./SearchBox.module.css";

// here we are going to specify the type of the parameters that
// SearchBox components have
interface ISearchBox {
  searchChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  clearMovies: () => void;
  handleSubmit: (searchField: string) => void;
  clearSearchField: () => void;
  searchField: string;
  error: string;
}

const SearchBox: React.FC<ISearchBox> = ({
  searchChange,
  handleSubmit,
  clearMovies,
  searchField,
  error,
  clearSearchField,
}) => {
  return (
    <div className={styles.container}>
      <div className={styles.searchContainer}>
      <input
        className={styles.input}
        type="search"
        placeholder="search movies..."
        onChange={searchChange}
        value={searchField}
      />

      <button
        className={styles.button}
        onClick={() => handleSubmit(searchField)}
      >
        Search
      </button>
      {searchField.length > 0 || error ? (
        <span
          className={styles.icon}
          role="img"
          aria-label="sheep"
          onClick={() => {
            clearMovies();
            clearSearchField();
          }}
        >
          &#10060;
        </span>
      ) : null}
      </div>
    </div>
  );
};

export default SearchBox;
