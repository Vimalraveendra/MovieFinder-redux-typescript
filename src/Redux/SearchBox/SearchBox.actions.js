import SearchBoxActionTypes from "./SearchBox.types";

export const searchChange = (film) => ({
  type: SearchBoxActionTypes.HANDLE_CHANGE,
  payload: film,
});
