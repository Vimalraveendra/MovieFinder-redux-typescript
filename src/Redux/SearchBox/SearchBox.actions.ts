import {
  HANDLE_SEARCH_CHANGE,
  CLEAR_SEARCH_FIELD,
  SearchBoxActionTypes,
} from "./SearchBox.types";

// TypeScript infers that this function is returning SearchBoxAction.
// in here we need to do two things  one is the type of the parameter
// and the  second one is to specify return type for each one of these actions
//(in our case it is the searcnBoxActionTypes ie SEARCHFIELD_CHANGE)
// so here searchBoxActionTypes  specify  all the possible actions
// that redux store can take place
export const searchChange = (
  film: React.ChangeEvent<HTMLInputElement>
): SearchBoxActionTypes => ({
  type: HANDLE_SEARCH_CHANGE,
  payload: film,
});

export const clearSearchField = (): SearchBoxActionTypes => ({
  type: CLEAR_SEARCH_FIELD,
});
