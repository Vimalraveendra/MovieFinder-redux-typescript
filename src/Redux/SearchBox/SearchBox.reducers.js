import SearchBoxActionTypes from "./SearchBox.types";

const initialState = {
  searchField: "",
};

export const searchChangeReducer = (state = initialState, action = {}) => {
  switch (action.type) {
    case SearchBoxActionTypes.HANDLE_SEARCH_CHANGE:
      return {
        ...state,
        searchField: action.payload,
      };
    case SearchBoxActionTypes.CLEAR_SEARCH_FIELD:
      return {
        ...state,
        searchField: action.payload,
      };
    default:
      return state;
  }
};
