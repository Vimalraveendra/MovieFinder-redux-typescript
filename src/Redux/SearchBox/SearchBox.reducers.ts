import {
  SearchBoxActionTypes,
  HANDLE_SEARCH_CHANGE,
  CLEAR_SEARCH_FIELD,
  ISearchBox,
} from "./SearchBox.types";

// the first thing to  do  specify type for the initial state or the default state.
// you have to make sure to import the types from the searchBox types.
const initialState: ISearchBox = {
  searchField: "",
};

// now we are going to specify the types for  parameters & return reducer types.
//Here we need to do two things , the first thing is to do a type
// for the action for that we need to import searchBoxActionsTypes.
// second thing is to do specify type for the return type as well
// in here we return the searchField  so here we need to say ISearchValue.
export const searchChangeReducer = (
  state = initialState,
  action: SearchBoxActionTypes
): ISearchBox => {
  switch (action.type) {
    case HANDLE_SEARCH_CHANGE:
      return {
        ...state,
        searchField: action.payload,
      };
    case CLEAR_SEARCH_FIELD:
      return {
        ...state,
        searchField: "",
      };
    default:
      return state;
  }
};
