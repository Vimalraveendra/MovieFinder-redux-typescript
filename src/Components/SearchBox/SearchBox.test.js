import { shallow, mount } from "enzyme";
import React from "react";
import SearchBox from "./SearchBox";
import { Provider } from "react-redux";

// import store from "../../Redux/store";
import configureStore from "redux-mock-store";
import { searchChange } from "../../Redux/SearchBox/SearchBox.actions";
const mockStore = configureStore();

describe("searchBox", () => {
  let wrapper, store;
  beforeEach(() => {
    const initialState = {
      searchText: { searchField: "" },
      moviesList: { error: "" },
    };

    store = mockStore(initialState);
    wrapper = shallow(<SearchBox store={store} />).dive();
  });
  it("should render the searchBox component", () => {
    // const initialState = {
    //   searchText: { searchField: "" },
    //   moviesList: { error: "" },
    // };
    // const store = mockStore(initialState);
    // const wrapper = mount(
    //   <Provider store={store}>
    //     <SearchBox />
    //   </Provider>
    // );
    expect(wrapper).toMatchSnapshot();
  });

  it("should  show the initial searchField value", () => {
    // test that the state values were correctly passed as props
    expect(wrapper.props().searchField).toEqual("");
    // const dispatch = jest.fn();
    // store.dispatch = dispatch;
  });

  it("should change the initial searchField value when ever the user types", () => {
    const dispatch = jest.fn();
    store.dispatch = dispatch;
    console.log(dispatch);

    expect(dispatch).toEqual("hai");
  });
});
