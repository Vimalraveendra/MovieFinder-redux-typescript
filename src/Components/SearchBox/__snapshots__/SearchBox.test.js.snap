// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`searchBox should render the searchBox component 1`] = `ShallowWrapper {}`;
